File: frontend/src/app/service/product/product.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {SaveProductRequestModel} from '../../model/save-product-request.model';
import {SaveProductResponse} from '../../interface/save-product-response';
import {Observable} from 'rxjs';
import {AllProductResponse} from '../../interface/all-product-response';
import {UpdateProductRequestModel} from '../../model/update-product-request.model';
import {UpdateProductResponse} from '../../interface/update-product-response';
import {SingleProductResponse} from '../../interface/single-product-response';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  http = inject(HttpClient);

  baseUrl = 'http://localhost:8080/product';

  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem('user') ? JSON.parse(localStorage.getItem('user')!).token : null;
    let headers = new HttpHeaders();
    if (token) {
      headers = headers.set('Authorization', `Bearer ${token}`);
    }
    return headers;
  }



  addProduct(payload: SaveProductRequestModel, files: File[]) : Observable<SaveProductResponse> {

    const formData = new FormData();
    formData.append('name', payload.name);
    formData.append('description', payload.description);
    formData.append('price', String(payload.price));
    formData.append('quantity', String(payload.quantity));
    formData.append("category", String(payload.category))

    files.forEach(file => formData.append('image', file))

    return this.http.post<SaveProductResponse>(`${this.baseUrl}/create`, formData, {headers: this.getAuthHeaders()});
  }

  getAllProducts() {
    return this.http.get<AllProductResponse[]>(`${this.baseUrl}/all`);
  }

  updateProduct(payload: UpdateProductRequestModel, id: number) {
    return this.http.put<UpdateProductResponse>(`${this.baseUrl}/update/${id}`, payload)
  }

  getProductById(id: number) {
    return this.http.get<SingleProductResponse>(`${this.baseUrl}/${id}`);
  }

  deleteProduct(id: number): Observable<string> {
    return this.http.delete<string>(`${this.baseUrl}/delete/${id}`, {
      headers: this.getAuthHeaders(),
      responseType: "text" as "json"
    });
  }


}

File: backend/src/main/java/am/itspace/backend/service/ProductService.java
package am.itspace.backend.service;

import am.itspace.backend.dto.AllProductResponse;
import am.itspace.backend.dto.ProductByIdResponse;
import am.itspace.backend.dto.SaveProductRequest;
import am.itspace.backend.dto.SaveProductResponse;
import am.itspace.backend.dto.UpdateProductRequest;
import am.itspace.backend.dto.UpdateProductResponse;
import am.itspace.backend.entity.Product;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.List;
import java.util.Optional;

public interface ProductService {

  SaveProductResponse create(SaveProductRequest request) throws IOException;

  Optional<ProductByIdResponse> getProductById(Long id);

  List<AllProductResponse> getAllProducts();

  UpdateProductResponse updateProduct(UpdateProductRequest request, Long id);

  void deleteProduct(Long id);

}
